sdk/cosmos/azure-cosmos/azure/cosmos/_base.py

```diff
@@ -674,3 +676,53 @@ def validate_cache_staleness_value(max_integrated_cache_staleness):
     if max_integrated_cache_staleness < 0:
         raise ValueError("Parameter 'max_integrated_cache_staleness_in_ms' can only be an "
                          "integer greater than or equal to zero")
+
+
+def _stringify_auto_scale(offer):
+    auto_scale_params = None
+    max_throughput = offer.auto_scale_max_throughput
+    increment_percent = offer.auto_scale_increment_percent
+    if max_throughput is not None:
+        auto_scale_params = {"maxThroughput": max_throughput}
+        if increment_percent is not None:
+            auto_scale_params["autoUpgradePolicy"] = {"throughputPolicy": {"incrementPercent": increment_percent}}
+    auto_scale_settings = json.dumps(auto_scale_params)
+
+    return auto_scale_settings
+
+
+def _set_throughput_options(offer, options):
+    offer_throughput = offer
+    request_options = options
+    if offer_throughput is not None:
+        try:
+            if offer_throughput.auto_scale_max_throughput:
+                request_options['autoUpgradePolicy'] = _stringify_auto_scale(offer=offer_throughput)
+            elif offer_throughput.auto_scale_increment_percent:
+                raise ValueError("auto_scale_max_throughput must be supplied in "
+                                 "conjunction with auto_scale_increment_percent")
+            elif offer_throughput.offer_throughput:
+                request_options["offerThroughput"] = offer_throughput.offer_throughput
+
+        except AttributeError:
+            if isinstance(offer_throughput, int):
+                request_options["offerThroughput"] = offer_throughput
```

Comment:
I would write it something like this to be clearer. You should add comments to indicate edge cases and so forth. 

```Python
if offer is None:
  return

if isinstance(offer, int):
  # do the int case here...
  return
elif hasattr(offer, "offer_throughput"):
  # the offer.offer_throughput here...
  return

# then this is the "normal" case that has validation
max_throughput = offer.auto_scale_max_throughput
increment_percent  = offer.auto_scale_increment_percent
if not(max_throughput and increment_percent):
  raise ValueError("...")
else:
  options['autoUpgradePolicy'] = _stringify_auto_scale(offer=offer)
```

Reply:
Whatever the logic actually should, the general thrust is to make it more readable and self-documenting.

Reply:
Okay understood, will take a deeper look into it.

Reply:
I did my best to make it more readable let me know what you think!